### PR TITLE
[dev] [DeepSeek-v4] Packed Sequence (THD) support for DSv4 Hybrid Attention

### PR DESCRIPTION
**Pull request by hxbai**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?
Add Packed Sequence (THD) support for DSv4 Hybrid Attention.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/transformer_layer.py [hunk 1]
```diff
@@ -1410,11 +1410,14 @@ def _te_cuda_graph_replay(self, *args, **kwargs):
             self.config.cuda_graph_modules
             and CudaGraphModule.attn not in self.config.cuda_graph_modules
         ):
+            input_ids = kwargs.get("input_ids", None)
             hidden_states, context = self._forward_attention(*args, **kwargs)
             args = (hidden_states,)
             kwargs = {}
             if padding_mask is not None:
                 kwargs["padding_mask"] = padding_mask
+            if input_ids is not None:
```
Comment (kunlunl) [inline megatron/core/transformer/transformer_layer.py:1419]: What are these lines and above line doing?

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 2]
```diff
@@ -815,33 +1217,71 @@ def forward(
...
+                    precomputed_grad_k_indexer = ig["d_index_k"].permute(1, 0, 2).contiguous()
+                    precomputed_grad_weights = ig["d_weights"].permute(1, 0, 2).contiguous()
             else:
                 attn_score_for_bwd = attn_score.clone()
                 index_score_for_bwd = index_score.clone()
+                dense_bwd_kwargs = {}
+                if is_thd:
+                    dense_bwd_kwargs = dict(
+                        cu_seqlens_q=cu_seqlens_q, cu_seqlens_k=cu_seqlens_compressed_idx
+                    )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py:1312]: fixed

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 3]
```diff
@@ -815,33 +1217,71 @@ def forward(
...
                     sm_scale=indexer_softmax_scale,
                     loss_coeff=indexer_loss_coeff,
                     grad_loss=unit_grad_loss,
                     block_I=128,
                 )
+
+                if is_thd:
+                    precomputed_grad_q_indexer = ig["d_index_q"].view(total_q, idx_nh, idx_hd)
+                    precomputed_grad_k_indexer = ig["d_index_k"].view(total_comp_idx, idx_hd)
+                    precomputed_grad_weights = ig["d_weights"].view(total_q, idx_nh)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: modified

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 4]
```diff
@@ -84,11 +84,349 @@ def get_compress_topk_idxs(
...
+        )
+        batch_of_comp = batch_of_row(cu_seqlens_compressed, total_q=total_comp_capacity)
+        valid_comp = src_comp < cu_seqlens_compressed[-1]
+        dst_comp = (
+            cu_seqlens_kv_full[batch_of_comp]
+            + kv_lens[batch_of_comp]
+            + (src_comp - cu_seqlens_compressed[batch_of_comp])
+        )
+        dst_comp = torch.where(valid_comp, dst_comp, total_kv + src_comp)
+        out[dst_comp] = compressed_kv_thd
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:333]: Added a comment to clarify

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 5]
```diff
@@ -979,3 +1623,521 @@ def forward(
...
+                "CompressedSparseAttention THD Path B requires at least "
+                "one segment with compressed indexer K; got none. (Should "
+                "be unreachable when ``n_compressed_total > 0``.)"
+            )
+
+        q_thd = q_indexer.squeeze(1)
+        w_thd = weights_indexer.squeeze(1)
+        k_thd = k_indexer.squeeze(1)
+
+        max_seqlen_compressed_idx = max_seqlen_q // self.compress_ratio
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: modified

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -677,8 +757,262 @@ def bwd_fused_indexer_loss_naive(
...
+            calculate_per_token_loss,
+        )
+        # topk_indices_b: (1, seqlen_q_b, topk_seg) where
+        # ``topk_seg = min(topk, seqlen_k_b)``. Real segments with
+        # ``seqlen_k_b < topk`` produce a narrower slice; write only
+        # those columns and leave the trailing ``[topk_seg:topk]``
+        # range at the buffer's initial -1 sentinel so the downstream
+        # post-filter in csa.py marks them invalid.
+        topk_seg = topk_indices_b.shape[-1]
+        topk_indices_thd[q_start:q_end, :topk_seg] = topk_indices_b.squeeze(0).int()
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:869]: Only fwd and bwd here do this. I'll leave it unmodified.

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 7]
```diff
@@ -592,6 +607,7 @@ def get_query_key_value_tensors(
             # In Megatron-Core, the qkv shape is [t, 1, h, d].
             # So we need to reshape qkv from [t, 1, h, d] to [t, h, d].
             q_compressed = q_compressed.squeeze(1)
+            kv_compressed = kv_compressed.squeeze(1)
 
         # =========================================
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py:612]: I'm sure it is correct.

File: megatron/core/transformer/moe/router.py [hunk 8]
```diff
@@ -714,7 +714,7 @@ def _apply_expert_bias(
         if self.enable_expert_bias and torch.is_grad_enabled():
             with torch.no_grad():
                 if padding_mask is not None:
-                    routing_map = routing_map & (~padding_mask)
+                    routing_map = routing_map & (~padding_mask.reshape(-1)).unsqueeze(-1)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/moe/router.py]: modified

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 9]
```diff
@@ -979,3 +1623,521 @@ def forward(
...
+                        self.softmax_scale,
+                        min(self.indexer.index_topk, max_seqlen_compressed_idx),
+                        indexer_loss_coeff,
+                        None,
+                        getattr(self.config, "dsa_indexer_use_sparse_loss", True),
+                        self.indexer.pg_collection,
+                        self.config.calculate_per_token_loss,
+                        cu_seqlens_q_for_loss,
+                        cu_seqlens_compressed_idx_for_loss,
+                        self.compress_ratio,
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:1749]: Added a comment to clarify

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 10]
```diff
@@ -815,33 +1217,71 @@ def forward(
...
+                        batch_size=-1,
+                        cu_seqlens_q=cu_seqlens_q,
+                        cu_seqlens_kv=cu_seqlens_compressed_idx,
+                    )
+                    bwd_q = q_indexer_flat.view(1, total_q, idx_nh, idx_hd)
+                    bwd_w = w_indexer.view(1, total_q, idx_nh)
+                    bwd_k = k_indexer_flat.view(1, total_comp_idx, idx_hd)
+                    bwd_attn = attn_score_for_bwd.view(1, total_q, topk)
+                    bwd_idx = index_score_for_bwd.view(1, total_q, topk)
+                    bwd_topk = topk_indices_cmp_global.view(1, total_q, topk)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: modified

File: megatron/core/transformer/multi_token_prediction.py [hunk 11]
```diff
@@ -2071,14 +2078,36 @@ def forward(
...
+                    packed_seq_params=packed_seq_params,
+                )
+                mtp_padding_mask = mtp_loss_mask == 0
+            else:
+                mtp_padding_mask = padding_mask
+
             layer_idx = 0 if self.mtp_use_repeated_layer else iteration
             (hidden_states, input_ids, position_ids, padding_mask) = self.layers[layer_idx](
                 input_ids=input_ids,
                 position_ids=position_ids,
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/multi_token_prediction.py]: modified for 2; added a CP TODO for 1

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 12]
```diff
@@ -979,3 +1623,521 @@ def forward(
...
+                cu_seqlens_kv_full,
+                cu_seqlens_compressed,
+                window_idxs,
+                max_seqlen_q,
+                packed_seq_params,
+            )
+        elif has_indexer and is_training:
+            output, indexer_loss = self._forward_fused_indexer_training_thd(
+                query,
+                x,
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:2140]: I think there is no need to do this simplification.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 13]
```diff
@@ -414,26 +414,106 @@ def fused_qk_topk_naive(
...
+        mask_b = _build_causal_mask_seg(sq_b, sk_b, ratio, device)
+
+        _, topk_b = fused_qk_topk_naive(q_b, k_b, w_b, index_topk, mask_b)
+        # topk_b: (1, sq_b, topk_k) where topk_k = min(index_topk, sk_b).
+        topk_b = topk_b.squeeze(0)
+        topk_k = topk_b.shape[-1]
+
+        # Mark invalid tail positions per row as ``-1``. A row at
+        # position ``i`` (0-indexed within the segment) has at most
+        # ``(i+1) // ratio`` causally-valid compressed positions; any
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:500]: It's a legacy name from previous PRs. I won't change it now.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 14]
```diff
@@ -407,9 +792,165 @@ def forward(self, x: torch.Tensor) -> Optional[torch.Tensor]:
...
+        offsets = torch.arange(ratio, device=device, dtype=base.dtype).unsqueeze(0)
+        gather_idx = base + offsets  # (total_comp, ratio)
+
+        kv_grouped = kv[gather_idx]  # (total_comp, ratio, 1, coff * d)
+        score_grouped = score[gather_idx]
+
+        # APE: (ratio, coff * d) → broadcast (1, ratio, 1, coff * d).
+        score_grouped = score_grouped + self.ape.view(1, ratio, 1, -1)
+
+        if self.overlap:
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:892]: This is an intermediate calculation. There is no need to add a helper.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/672ddcc7045487a45fb752f023ffe9d6280fb522

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/dd64c9115fcd5ab1c2972bfdeaee6c448ec0a7e4

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/8e4bfd3f29badce445bc323ec4c1303ef7b1dc41

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/d7549101a98521586e61cc136b2fd6f003c9f485

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/1c33ca14a133d328ff79a5112a7dfdf73e24bab4

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/a41693bbd8efc5fce290d3ddd1532b5e856ef78f

**Review by kunlunl — APPROVED** [review]

File: megatron/core/transformer/transformer_layer.py (hunk 1 above)
Comment (hxbai) [reply megatron/core/transformer/transformer_layer.py:1419]: They are for HashMoE with CUDAGraph. In some partial cudagraph cases, this input is missing. So, I added it here.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/7f3c63383b7839163a4e6b0c2d14b2c4115b39e1

**Review by yaox12 — APPROVED** [review]

File: megatron/core/transformer/multi_token_prediction.py [hunk 15]
```diff
@@ -2071,14 +2078,40 @@ def forward(
...
                 hidden_states=hidden_states,
                 attention_mask=attention_mask,
-                padding_mask=padding_mask,
+                padding_mask=mtp_padding_mask,
```
Comment (Victarry) [inline megatron/core/transformer/multi_token_prediction.py]: **[IMPORTANT Correctness]** `mtp_padding_mask` is derived from an already-rolled `mtp_loss_mask`, then `MultiTokenPredictionLayer._get_embeddings()` rolls the provided `padding_mask` again before routing. That makes the first MTP MoE layer receive a mask shifted by two tokens while `input_ids`/`position_ids` are shifted once, so valid tokens can be excluded and boundary/padding tokens can enter router aux/z/bias statistics.

**Suggestion:** Pass the unshifted padding mask into the layer and let `_get_embeddings()` roll it exactly once, or derive the padding mask inside the layer after the input roll so the token ids, positions, and routing mask stay aligned.

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 16]
```diff
@@ -681,101 +956,214 @@ class FusedIndexerSparseAttnFunc(torch.autograd.Function):
...
-        n_comp = k_indexer.shape[0]
-        idx_nh, idx_hd = q_indexer.shape[2], q_indexer.shape[3]
+        is_thd = cu_seqlens_q is not None
 
-        effective_topk = min(indexer_topk, n_comp)
+        # ---- Layout-specific input prep --------------------------------------
+        # SBHD: permute SBHD→BSHD once and reuse the BSHD tensors for indexer
+        # forward, dense score helpers, and the indexer backward.
+        # THD: skip the permute; tensors are already flat.
+        if is_thd:
```
Comment (Victarry) [inline megatron/core/transformer/experimental_attention_variant/dsa_kernels.py:1034]: **[IMPORTANT Correctness]** In the fused THD Path B indexer-loss path, CUDA-graph padding rows are still treated as real query rows. `total_q` comes from the padded `q_indexer` tensor, and `_indexer_topk_core` uses the padded `cu_seqlens_q`, so dummy/tail rows with at least one causally valid compressed KV can produce indexer scores, target/predict distributions, and KL terms. The loss reducers only mask rows with no valid top-k/KV; they do not consult `loss_mask` or `padding_mask`. The unfused THD path rebuilds unpadded cu_seqlens for the loss path, which suggests the fused path should do the same.

**Suggestion:** Use true unpadded `cu_seqlens` or pass an explicit valid-row mask for the indexer loss/backward path, so padding rows keep the static attention shape but do not contribute to `DSAIndexerLossAutoScaler` gradients.

File: megatron/core/transformer/multi_token_prediction.py (hunk 15 above)
Comment (hxbai) [reply megatron/core/transformer/multi_token_prediction.py]: removed the new added part, and fixed the previous wrong roll in _get_embeddings

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py (hunk 16 above)
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py:1034]: fixed

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/215de5e725f09a7d08ad6a1f7180c40f2777b242

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/7e38374a91abdbb5fb74f05ebb272bb1a704312b

**Review by Victarry — APPROVED** [review]

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5011/commits/e1bdcccfffd3718d7a51f0d24893b6f475c5fd65
